dbms/src/Storages/MergeTree/MergeTreeRangeReader.cpp

```diff
@@ -581,44 +601,46 @@ MergeTreeRangeReader::ReadResult MergeTreeRangeReader::startReadingChain(size_t
         {
             if (stream.isFinished())
             {
-                result.addRows(stream.finalize(result.block));
+                result.addRows(stream.finalize(result.columns));
                 stream = Stream(ranges.back().begin, ranges.back().end, merge_tree_reader);
                 result.addRange(ranges.back());
                 ranges.pop_back();
             }
 
             auto rows_to_read = std::min(space_left, stream.numPendingRowsInCurrentGranule());
             bool last = rows_to_read == space_left;
-            result.addRows(stream.read(result.block, rows_to_read, !last));
+            result.addRows(stream.read(result.columns, rows_to_read, !last));
             result.addGranule(rows_to_read);
             space_left -= rows_to_read;
         }
     }
 
-    result.addRows(stream.finalize(result.block));
+    result.addRows(stream.finalize(result.columns));
 
     /// Last granule may be incomplete.
     result.adjustLastGranule();
 
     return result;
 }
 
-Block MergeTreeRangeReader::continueReadingChain(ReadResult & result)
+Columns MergeTreeRangeReader::continueReadingChain(ReadResult & result, size_t & num_rows)
 {
-    Block block;
+    Columns columns;
+    num_rows = 0;
 
     if (result.rowsPerGranule().empty())
     {
         /// If zero rows were read on prev step, than there is no more rows to read.
         /// Last granule may have less rows than index_granularity, so finish reading manually.
         stream.finish();
-        return block;
+        return columns;
     }
 
+    columns.resize(merge_tree_reader->getColumns().size());
```

Comment:
Slightly confusing. Previously we haven't got `Columns`, and `getColumns()` method had looked definitely.